onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -101,37 +105,79 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
-      return false;
-    }
-
-    const Node& unsqueeze = edges[0]->GetNode();
-    const Node& gather = edges[1]->GetNode();
-    const Node& shape = edges[2]->GetNode();
-
-    const NodeArg& shape_input = *(shape.InputDefs()[0]);
-    if (shape_input.Name() != root_input.Name()) {
-      return false;
-    }
-
-    std::vector<int64_t> axes;
-    if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
-      return false;
+    if (graph_utils::FindPath(concat, true, parent_path, edges, logger)) {
+      const Node& unsqueeze = edges[0]->GetNode();
+      const Node& gather = edges[1]->GetNode();
+      const Node& shape = edges[2]->GetNode();
+
+      const NodeArg& shape_input = *(shape.InputDefs()[0]);
+      if (shape_input.Name() != root_input.Name()) {
+        return false;
+      }
+
+      std::vector<int64_t> axes;
+      if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
+        return false;
+      }
+
+      if (!optimizer_utils::IsInitializerWithExpectedValue(graph, *(gather.InputDefs()[1]), int64_t(shape_value.size()), false)) {
+        return false;
+      }
+
+      shape_value.push_back(0);
+
+      candidates_for_removal.push_back(unsqueeze.Index());
+      candidates_for_removal.push_back(gather.Index());
+      candidates_for_removal.push_back(shape.Index()); 
+    } else {
+      // Find subgraph
+      if (subgraph_cnt > 0 || std::find(shape_value.begin(), shape_value.end(), -1) != shape_value.end()) {
+        // Only one subgraph can be used to fuse, and each shape node should contain only one value of -1. 
+        return false;
+      }
+      Node* p_cur_node = const_cast<Node*>(graph_utils::GetInputNode(concat, i));
+      if (p_cur_node == nullptr) {
+        return false;
+      }
+      subgraph_candidates.clear();
+      // From the current node, find the subgraph bottom-up util it reaches the root input node. 
+      while (p_cur_node != nullptr) {
+        Node& cur_node = *p_cur_node;
+        // Each eligible node in the subgraph must have exactly one output node and less than one 
+        // input edge(the node input could be graph input). 
+        if (cur_node.GetOutputEdgesCount() != 1 || cur_node.GetInputEdgesCount() > 1) {
+          break;
+        }
+        // Match input of current node with root node.
+        bool input_matched = false;
+        for (int j = 0; j < cur_node.InputDefs().size(); j++) {
+          if (graph_utils::GetNodeInputName(cur_node, j) == root_input.Name()) {
+            input_matched = true;
+            break;
+          }
+        }
+        if (!input_matched) {
+          // No input in current node matches root input. Go to the next node.
+          subgraph_candidates.push_back(cur_node.Index());
+          p_cur_node = const_cast<Node*>(graph_utils::GetInputNode(cur_node, 0));
```

Comment:
The index of input is not alawys 0. Current node might have several inputs. Probably some inputs are initializers or constant.